Userland/Applications/Mail/MailWidget.cpp

```diff
@@ -0,0 +1,480 @@
+        return false;
+    }
+
+    m_imap_client = adopt_own(*new IMAP::Client(server, port, tls));
+    auto connection_promise = m_imap_client->connect();
+    if (!connection_promise.has_value()) {
+        GUI::MessageBox::show_error(window(), String::formatted("Failed to connect to '{}:{}' over {}.", server, port, tls ? "TLS" : "Plaintext"));
+        return false;
+    }
+    connection_promise.value()->await();
+
+    auto response = m_imap_client->login(username, password)->await().release_value();
+
+    if (response.status() != IMAP::ResponseStatus::OK) {
+        dbgln("Failed to login. The server says: '{}'", response.response_text());
+        GUI::MessageBox::show_error(window(), String::formatted("Failed to login. The server says: '{}'", response.response_text()));
+        return false;
+    }
+
+    response = m_imap_client->list("", "*")->await().release_value();
+
+    if (response.status() != IMAP::ResponseStatus::OK) {
+        dbgln("Failed to retrieve mailboxes. The server says: '{}'", response.response_text());
+        GUI::MessageBox::show_error(window(), String::formatted("Failed to retrieve mailboxes. The server says: '{}'", response.response_text()));
+        return false;
+    }
+
+    auto& list_items = response.data().list_items();
+
+    m_account_holder = AccountHolder::create();
+    m_account_holder->add_account_with_name_and_mailboxes(username, move(list_items));
+
+    m_mailbox_list->set_model(m_account_holder->mailbox_tree_model());
+    m_mailbox_list->expand_tree();
+
+    return true;
+}
+
+void MailWidget::on_window_close()
+{
+    auto response = move(m_imap_client->send_simple_command(IMAP::CommandType::Logout)->await().release_value().get<IMAP::SolidResponse>());
+    VERIFY(response.status() == IMAP::ResponseStatus::OK);
+
+    m_imap_client->close();
+}
+
+IMAP::MultiPartBodyStructureData const* MailWidget::look_for_alternative_body_structure(IMAP::MultiPartBodyStructureData const& current_body_structure, Vector<int>& position_stack) const
+{
+    if (current_body_structure.media_type.equals_ignoring_case("ALTERNATIVE"))
+        return &current_body_structure;
+
+    int structure_index = 1;
+
+    for (auto& structure : current_body_structure.bodies) {
+        if (structure->data().has<IMAP::BodyStructureData>()) {
+            ++structure_index;
+            continue;
+        }
+
+        position_stack.append(structure_index);
+        auto* potential_alternative_structure = look_for_alternative_body_structure(structure->data().get<IMAP::MultiPartBodyStructureData>(), position_stack);
+
+        if (potential_alternative_structure)
+            return potential_alternative_structure;
+
+        position_stack.take_last();
+        ++structure_index;
+    }
+
+    return nullptr;
+}
+
+Vector<MailWidget::Alternative> MailWidget::get_alternatives(IMAP::MultiPartBodyStructureData const& multi_part_body_structure_data) const
+{
+    Vector<int> position_stack;
```

Comment:
This should probably be `size_t` or some other unsigned type.

Reply:
The IMAP library currently uses `int` instead of `size_t`, so I went with `int` to avoid casting.
Should I change the IMAP library to use unsigned types in this PR?